src/Common/FitOnFhir.Common/Services/AuthStateService.cs

```diff
@@ -79,7 +105,74 @@ public AuthState CreateAuthState(HttpRequest httpRequest)
                 ExternalSystem = jwtSecurityToken.Issuer;
             }
 
-            return new AuthState(ExternalIdentifier, ExternalSystem);
+            var redirectUrl = HttpUtility.UrlDecode(httpRequest.Query[Constants.RedirectUrlQueryParameter]);
+            if (_authenticationConfiguration.ApprovedRedirectUrls.All(url => url != redirectUrl))
+            {
+                string errorMessage = $"{redirectUrl} was not found in the list of approved redirect URLs.";
+                _logger.LogError(errorMessage);
+                throw new RedirectUrlException(errorMessage);
+            }
+
+            RedirectUrl = redirectUrl;
+            State = HttpUtility.UrlDecode(httpRequest.Query[Constants.StateQueryParameter]);
+
+            return new AuthState(ExternalIdentifier, ExternalSystem, RedirectUrl, State);
+        }
+
+        public async Task<AuthState> RetrieveAuthState(string nonce, CancellationToken cancellationToken)
+        {
+            try
+            {
+                var blobName = EnsureArg.IsNotNullOrWhiteSpace(nonce);
+
+                // Get a reference to the blob
+                BlobClient blobClient = _blobContainerClient.GetBlobClient(blobName);
+
+                // Get the AuthState
+                var response = await blobClient.DownloadAsync(cancellationToken);
+
+                // deserialize the AuthState
+                StreamReader reader = new StreamReader(response.Value.Content);
```

Comment:
Is there a chance that `response.Value.Content` can be null?

Reply:
I think the only way that can happen is if somehow a blob was uploaded with nothing in it.  The blob is guaranteed to exist, because we are getting a client to it from the nonce handle.  I can't imagine a scenario where a blob is created via `StoreAuthState` but nothing is uploaded with it.